File: src/main/java/me/m64diamondstar/ingeniamccore/General/Commands/Ingenia/BalanceSubcommand.java
package me.m64diamondstar.ingeniamccore.General.Commands.Ingenia;

import me.m64diamondstar.ingeniamccore.General.Player.IngeniaPlayer;
import me.m64diamondstar.ingeniamccore.Utils.Colors;
import me.m64diamondstar.ingeniamccore.Utils.MessageType;
import me.m64diamondstar.ingeniamccore.Utils.Messages;
import org.bukkit.Bukkit;
import org.bukkit.command.CommandSender;

public class BalanceSubcommand {

    private CommandSender sender;
    private String[] args;

    public BalanceSubcommand(CommandSender sender, String[] args){
        this.sender = sender;
        this.args = args;
    }

    /**
     * Execute the command
     */
    public void execute(){
        if(args.length <= 2){
            sender.sendMessage(Colors.format(Messages.commandUsage("ig balance <add/set/subtract/get> <player> [amount]")));
        }

        else if(args.length == 3){

            if(args[1].equalsIgnoreCase("get")){
                IngeniaPlayer target = new IngeniaPlayer(Bukkit.getPlayer(args[2]));
                if(target.getPlayer() == null){
                    sender.sendMessage(Messages.invalidPlayer());
                    return;
                }
                sender.sendMessage(Colors.format(target.getName() + " has " + target.getBal() + ":gs:.", MessageType.INFO));
            }else{
                sender.sendMessage(Colors.format(Messages.commandUsage("ig balance <add/set/subtract/get> <player> [amount]")));
            }

        }

        else if(args.length == 4){

            long gs;

            try{
                gs = Long.parseLong(args[3]);
            }catch (NumberFormatException e){
                sender.sendMessage(Messages.invalidNumber());
                return;
            }

            IngeniaPlayer target = new IngeniaPlayer(Bukkit.getPlayer(args[2]));
            if(target.getPlayer() == null){
                sender.sendMessage(Messages.invalidPlayer());
                return;
            }

            if(args[1].equalsIgnoreCase("add")){
                target.addBal(gs);
                sender.sendMessage(Colors.format("Successfully added " + gs + ":gs: to " + target.getName() + ".", MessageType.SUCCESS));
            }else if(args[1].equalsIgnoreCase("set")){
                target.setBal(gs);
                sender.sendMessage(Colors.format("Successfully set " + target.getName() + "'s balance to " + gs + ":gs:.", MessageType.SUCCESS));
            }else if(args[1].equalsIgnoreCase("subtract")){
                target.addBal(-gs);
                sender.sendMessage(Colors.format("Successfully subtracted " + gs + ":gs: from " + target.getName() + ".", MessageType.SUCCESS));
            }else{
                sender.sendMessage(Colors.format(Messages.commandUsage("ig balance <add/set/subtract/get> <player> [amount]")));
            }

        }
    }

}

File: src/main/java/me/m64diamondstar/ingeniamccore/general/commands/ingenia/BalanceSubcommand.kt
package me.m64diamondstar.ingeniamccore.general.commands.ingenia

import me.m64diamondstar.ingeniamccore.general.player.IngeniaPlayer
import me.m64diamondstar.ingeniamccore.utils.IngeniaSubcommand
import me.m64diamondstar.ingeniamccore.utils.messages.Colors
import me.m64diamondstar.ingeniamccore.utils.messages.MessageType
import me.m64diamondstar.ingeniamccore.utils.messages.Messages
import org.bukkit.Bukkit
import org.bukkit.command.CommandSender
import java.lang.NumberFormatException

class BalanceSubcommand(private val sender: CommandSender, private val args: Array<String>): IngeniaSubcommand {

    /**
     * Execute the command
     */
    override fun execute() {
        if (args.size <= 2) {
            sender.sendMessage(Colors.format(Messages.commandUsage("ig balance <add/set/subtract/get> <player> [amount]")))
        } else if (args.size == 3) {
            if (args[1].equals("get", ignoreCase = true)) {
                val target = Bukkit.getPlayer(args[2])
                if (target == null) {
                    sender.sendMessage(Messages.invalidPlayer())
                    return
                }
                val targetPlayer = IngeniaPlayer(target)
                sender.sendMessage(
                    Colors.format(
                        MessageType.INFO + targetPlayer.name + " has " + targetPlayer.bal + ":gs:."))
            } else {
                sender.sendMessage(
                    Colors.format(
                        Messages.commandUsage("ig balance <add/set/subtract/get> <player> [amount]")))
            }
        } else if (args.size == 4) {
            val gs: Long = try {
                args[3].toLong()
            } catch (e: NumberFormatException) {
                sender.sendMessage(Messages.invalidNumber())
                return
            }
            val target = Bukkit.getPlayer(args[2])
            if (target == null) {
                sender.sendMessage(Messages.invalidPlayer())
                return
            }
            val targetPlayer = IngeniaPlayer(target)
            if (args[1].equals("add", ignoreCase = true)) {
                targetPlayer.addBal(gs)
                sender.sendMessage(
                    Colors.format(
                        MessageType.SUCCESS + "Successfully added " + gs + ":gs: to " + targetPlayer.name + "."
                    )
                )
            } else if (args[1].equals("set", ignoreCase = true)) {
                targetPlayer.bal = gs
                sender.sendMessage(
                    Colors.format(
                        MessageType.SUCCESS + "Successfully set " + targetPlayer.name + "'s balance to " + gs + ":gs:."
                    )
                )
            } else if (args[1].equals("subtract", ignoreCase = true)) {
                targetPlayer.addBal(-gs)
                sender.sendMessage(
                    Colors.format(
                        MessageType.SUCCESS + "Successfully subtracted " + gs + ":gs: from " + targetPlayer.name + "."
                    )
                )
            } else {
                sender.sendMessage(
                    Colors.format(
                        Messages.commandUsage("ig balance <add/set/subtract/get> <player> [amount]")))
            }
        }
    }


    override fun getTabCompleters(): ArrayList<String>{
        var tabs = ArrayList<String>()

        val firstTab = ArrayList<String>()
        firstTab.add("add")
        firstTab.add("set")
        firstTab.add("subtract")
        firstTab.add("get")

        val secondTab = ArrayList<String>()
        for(player in Bukkit.getOnlinePlayers())
            secondTab.add(player.name)

        val thirdTab = ArrayList<String>()
        thirdTab.add("1")
        thirdTab.add("2")
        thirdTab.add("3")
        thirdTab.add("4")
        thirdTab.add("5")
        thirdTab.add("6")
        thirdTab.add("7")
        thirdTab.add("8")
        thirdTab.add("9")

        if(args.size == 2)
            tabs = firstTab
        else if(args.size == 3)
            tabs = secondTab
        else if(args.size == 4 && !args[1].equals("get", true))
            tabs = thirdTab

        return tabs
    }
}